src/passes/PostAssemblyScript.cpp

```diff
@@ -0,0 +1,530 @@
+
+  // Gets that are escapes
+  std::unordered_set<LocalGet*> escapes;
+
+  void visitLocalSet(LocalSet* curr) {
+    // local.set(X, __retain(...)) ?
+    if (auto* call = curr->value->dynCast<Call>()) {
+      if (call->target == RETAIN) {
+        retains[curr] = getCurrentPointer();
+      }
+    }
+  }
+
+  void visitCall(Call* curr) {
+    // __release(local.get(X, ...)) ?
+    if (curr->target == RELEASE && curr->operands.size() == 1) {
+      auto* operand = curr->operands[0];
+      if (auto* localGet = operand->dynCast<LocalGet>()) {
+        releases[localGet] = getCurrentPointer();
+      }
+    }
+  }
+
+  void visitReturn(Return* curr) {
+    // return(local.get(X, ...)) ?
+    auto* value = curr->value;
+    if (value) {
+      if (auto* localGet = value->dynCast<LocalGet>()) {
+        escapes.insert(localGet);
+      }
+    }
+  }
+
+  void visitThrow(Throw* curr) {
+    // throw(..., local.get(X, ...), ...) ?
+    for (auto* operand : curr->operands) {
+      if (auto* localGet = operand->dynCast<LocalGet>()) {
+        escapes.insert(localGet);
+        break;
+      }
+    }
+  }
+
+  void eliminateRetain(Expression** location) {
+    // assumes local.set(X, __retain(...))
+    auto* localSet = (*location)->cast<LocalSet>();
+    auto* call = localSet->value->cast<Call>();
+    assert(call->target == RETAIN);
+    assert(call->operands.size() == 1);
+    localSet->value = call->operands[0];
+  }
+
+  void eliminateRelease(Expression** location) {
+    // assumes __release(local.get(X, ...))
+    if (auto* call = (*location)->dynCast<Call>()) {
+      assert(call->target == RELEASE);
+      assert(call->operands.size() == 1);
+      assert(call->operands[0]->is<LocalGet>());
+      Builder builder(*getModule());
+      *location = builder.makeNop();
+    } else {
+      assert((*location)->is<Nop>()); // already replaced
+    }
+  }
+
+  // Tests if a retain reaches an escape and thus is
+  // considered necessary.
+  bool testReachesEscape(LocalSet* retain, AliasGraph& graph) {
+    for (auto* localGet : graph.setInfluences[retain]) {
+      if (releases.find(localGet) != releases.end()) {
+        continue;
+      }
+      if (escapes.find(localGet) != escapes.end()) {
+        return true;
+      }
+      for (auto* localSet : graph.getInfluences[localGet]) {
+        if (testReachesEscape(localSet, graph)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  // Collects all reachable releases of a retain
+  void collectReleases(LocalSet* retain,
+                       AliasGraph& graph,
+                       std::unordered_set<Expression**>& found) {
+    for (auto* localGet : graph.setInfluences[retain]) {
+      auto foundRelease = releases.find(localGet);
+      if (foundRelease != releases.end()) {
+        found.insert(foundRelease->second);
+      } else {
+        for (auto* localSet : graph.getInfluences[localGet]) {
+          collectReleases(localSet, graph, found);
```

Comment:
Same infinite recursion danger here, I think? In this case you'd also need to indicate an error if a recursion limit is hit, I guess.